File: backend/app/core/security/encryption.py
"""
Data-at-Rest Encryption Module
ISO 27001 A.10.1.1 - Policy on the use of cryptographic controls
ISO 27001 A.10.1.2 - Key management

Provides AES-256-GCM encryption for sensitive data stored in Redis and databases.
Implements NIST SP 800-38D authenticated encryption with additional data (AEAD).

@module core.security.encryption
"""

import os
import base64
import hashlib
import secrets
from typing import Tuple, Optional, Dict, Any
from datetime import datetime, timezone
from enum import Enum

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.exceptions import InvalidTag

from app.core.logging import get_logger, get_audit_logger
from app.core.logging.audit import AuditEventType, AuditSeverity

logger = get_logger(__name__)
audit_logger = get_audit_logger()


class EncryptionError(Exception):
    """Base exception for encryption errors."""
    pass


class DecryptionError(Exception):
    """Base exception for decryption errors."""
    pass


class KeyDerivationError(Exception):
    """Base exception for key derivation errors."""
    pass


class DataClassification(str, Enum):
    """
    Data classification levels (ISO 27001 A.8.2.1).

    Determines encryption requirements and key rotation policies.
    """
    PUBLIC = "public"              # No encryption required
    INTERNAL = "internal"          # Standard encryption
    CONFIDENTIAL = "confidential"  # Enhanced encryption + strict access
    PHI = "phi"                    # HIPAA PHI - highest security
    PII = "pii"                    # Personal Identifiable Information


# ============================================================================
# AES-256-GCM ENCRYPTION ENGINE
# ============================================================================

class AESGCMEncryption:
    """
    AES-256-GCM authenticated encryption.

    ISO 27001 A.10.1.1 - Uses NIST-approved algorithm
    NIST SP 800-38D - Galois/Counter Mode

    Features:
    - AES-256 encryption (256-bit keys)
    - GCM mode (Galois/Counter Mode)
    - AEAD (Authenticated Encryption with Associated Data)
    - Random 96-bit nonces (never reused)
    - 128-bit authentication tags
    """

    # Key size in bytes (256 bits = 32 bytes)
    KEY_SIZE = 32

    # Nonce size in bytes (96 bits = 12 bytes, recommended for GCM)
    NONCE_SIZE = 12

    # Tag size in bytes (128 bits = 16 bytes, maximum for GCM)
    TAG_SIZE = 16

    def __init__(self, key: bytes):
        """
        Initialize AES-GCM cipher.

        Args:
            key: 32-byte (256-bit) encryption key

        Raises:
            ValueError: If key size is invalid
        """
        if len(key) != self.KEY_SIZE:
            raise ValueError(f"Key must be exactly {self.KEY_SIZE} bytes (256 bits)")

        self.cipher = AESGCM(key)
        self.key_id = hashlib.sha256(key).hexdigest()[:16]  # Key fingerprint

        logger.debug(f"AES-GCM cipher initialized (key_id: {self.key_id})")

    def encrypt(
        self,
        plaintext: bytes,
        associated_data: Optional[bytes] = None
    ) -> Tuple[bytes, bytes]:
        """
        Encrypt plaintext using AES-256-GCM.

        Args:
            plaintext: Data to encrypt
            associated_data: Optional additional authenticated data (not encrypted)

        Returns:
            Tuple of (ciphertext, nonce)

        Raises:
            EncryptionError: If encryption fails
        """
        try:
            # Generate random nonce (MUST be unique for each encryption)
            nonce = os.urandom(self.NONCE_SIZE)

            # Encrypt with authentication
            ciphertext = self.cipher.encrypt(nonce, plaintext, associated_data)

            return ciphertext, nonce

        except Exception as e:
            logger.error(f"Encryption failed: {e}", exc_info=True)
            raise EncryptionError(f"Encryption failed: {e}")

    def decrypt(
        self,
        ciphertext: bytes,
        nonce: bytes,
        associated_data: Optional[bytes] = None
    ) -> bytes:
        """
        Decrypt ciphertext using AES-256-GCM.

        Args:
            ciphertext: Encrypted data (includes authentication tag)
            nonce: Nonce used during encryption
            associated_data: Optional additional authenticated data

        Returns:
            Decrypted plaintext

        Raises:
            DecryptionError: If decryption or authentication fails
        """
        try:
            # Decrypt and verify authentication tag
            plaintext = self.cipher.decrypt(nonce, ciphertext, associated_data)

            return plaintext

        except InvalidTag:
            # Authentication tag verification failed - data was tampered with
            audit_logger.log_security_event(
                event_type=AuditEventType.SECURITY_DATA_INTEGRITY_VIOLATION,
                severity=AuditSeverity.CRITICAL,
                description="Data integrity violation: Authentication tag verification failed",
                metadata={
                    'key_id': self.key_id,
                    'error': 'Invalid authentication tag'
                }
            )
            raise DecryptionError("Data integrity violation: Authentication failed")

        except Exception as e:
            logger.error(f"Decryption failed: {e}", exc_info=True)
            raise DecryptionError(f"Decryption failed: {e}")


# ============================================================================
# KEY DERIVATION (PBKDF2)
# ============================================================================

class KeyDerivation:
    """
    PBKDF2-HMAC-SHA256 key derivation.

    ISO 27001 A.10.1.2 - Key management
    NIST SP 800-132 - Recommendation for Password-Based Key Derivation
    """

    # Default iterations (100,000+ recommended by NIST)
    DEFAULT_ITERATIONS = 100_000

    # Salt size in bytes (128 bits minimum)
    SALT_SIZE = 16

    @staticmethod
    def derive_key(
        master_key: str,
        salt: bytes,
        iterations: int = DEFAULT_ITERATIONS,
        key_length: int = AESGCMEncryption.KEY_SIZE
    ) -> bytes:
        """
        Derive encryption key from master key using PBKDF2.

        Args:
            master_key: Master key (password/secret)
            salt: Random salt (must be stored)
            iterations: KDF iterations (higher = more secure, slower)
            key_length: Derived key length in bytes

        Returns:
            Derived key

        Raises:
            KeyDerivationError: If derivation fails
        """
        try:
            kdf = PBKDF2HMAC(
                algorithm=hashes.SHA256(),
                length=key_length,
                salt=salt,
                iterations=iterations,
            )

            derived_key = kdf.derive(master_key.encode('utf-8'))

            return derived_key

        except Exception as e:
            logger.error(f"Key derivation failed: {e}", exc_info=True)
            raise KeyDerivationError(f"Key derivation failed: {e}")

    @staticmethod
    def generate_salt() -> bytes:
        """Generate random salt for key derivation."""
        return os.urandom(KeyDerivation.SALT_SIZE)


# ============================================================================
# ENCRYPTION SERVICE (HIGH-LEVEL API)
# ============================================================================

class EncryptionService:
    """
    High-level encryption service for data-at-rest protection.

    ISO 27001 A.10.1.1, A.10.1.2
    HIPAA § 164.312(a)(2)(iv) - Encryption and decryption

    Usage:
        >>> service = EncryptionService(master_key="your-master-key")
        >>> encrypted, metadata = service.encrypt_data(
        ...     data=b"sensitive data",
        ...     classification=DataClassification.PHI
        ... )
        >>> decrypted = service.decrypt_data(encrypted, metadata)
    """

    def __init__(
        self,
        master_key: str,
        kdf_iterations: int = KeyDerivation.DEFAULT_ITERATIONS
    ):
        """
        Initialize encryption service.

        Args:
            master_key: Master encryption key (from environment)
            kdf_iterations: PBKDF2 iterations
        """
        self.master_key = master_key
        self.kdf_iterations = kdf_iterations

        # Derive salt deterministically from master key to ensure consistency
        # This ensures the same master key always produces the same derived key
        import hashlib
        salt_source = f"encryption-salt-v1-{master_key}".encode('utf-8')
        self.salt = hashlib.sha256(salt_source).digest()[:KeyDerivation.SALT_SIZE]

        # Derive encryption key from master key
        self.encryption_key = KeyDerivation.derive_key(
            master_key=master_key,
            salt=self.salt,
            iterations=kdf_iterations
        )

        # Initialize cipher
        self.cipher = AESGCMEncryption(self.encryption_key)

        logger.info(
            "Encryption service initialized",
            extra={
                "key_id": self.cipher.key_id,
                "kdf_iterations": kdf_iterations,
                "iso27001_control": "A.10.1.1, A.10.1.2"
            }
        )

    def encrypt_data(
        self,
        data: bytes,
        classification: DataClassification = DataClassification.CONFIDENTIAL,
        context: Optional[Dict[str, Any]] = None
    ) -> Tuple[bytes, Dict[str, Any]]:
        """
        Encrypt data with metadata.

        Args:
            data: Plaintext data to encrypt
            classification: Data classification level
            context: Optional context metadata (user_id, resource_id, etc.)

        Returns:
            Tuple of (encrypted_data, metadata)

        Raises:
            EncryptionError: If encryption fails
        """
        try:
            # Prepare associated data (authenticated but not encrypted)
            associated_data = self._build_associated_data(classification, context)

            # Encrypt
            ciphertext, nonce = self.cipher.encrypt(
                plaintext=data,
                associated_data=associated_data
            )

            # Build metadata for storage
            metadata = {
                'version': '1.0',
                'algorithm': 'AES-256-GCM',
                'key_id': self.cipher.key_id,
                'nonce': base64.b64encode(nonce).decode('utf-8'),
                'salt': base64.b64encode(self.salt).decode('utf-8'),
                'kdf_iterations': self.kdf_iterations,
                'classification': classification.value,
                'encrypted_at': datetime.now(timezone.utc).isoformat(),
                'context': context or {}
            }

            # Log encryption event
            logger.debug(
                f"Data encrypted successfully",
                extra={
                    'key_id': self.cipher.key_id,
                    'classification': classification.value,
                    'data_size': len(data)
                }
            )

            return ciphertext, metadata

        except Exception as e:
            logger.error(f"Data encryption failed: {e}", exc_info=True)
            raise EncryptionError(f"Data encryption failed: {e}")

    def decrypt_data(
        self,
        ciphertext: bytes,
        metadata: Dict[str, Any]
    ) -> bytes:
        """
        Decrypt data using stored metadata.

        Args:
            ciphertext: Encrypted data
            metadata: Encryption metadata (from encrypt_data)

        Returns:
            Decrypted plaintext

        Raises:
            DecryptionError: If decryption fails
        """
        try:
            # Validate metadata
            self._validate_metadata(metadata)

            # Extract nonce
            nonce = base64.b64decode(metadata['nonce'])

            # Rebuild associated data
            classification = DataClassification(metadata['classification'])
            context = metadata.get('context')
            associated_data = self._build_associated_data(classification, context)

            # Decrypt
            plaintext = self.cipher.decrypt(
                ciphertext=ciphertext,
                nonce=nonce,
                associated_data=associated_data
            )

            # Log decryption event
            logger.debug(
                f"Data decrypted successfully",
                extra={
                    'key_id': metadata['key_id'],
                    'classification': classification.value
                }
            )

            return plaintext

        except Exception as e:
            logger.error(f"Data decryption failed: {e}", exc_info=True)
            raise DecryptionError(f"Data decryption failed: {e}")

    def encrypt_string(
        self,
        plaintext: str,
        classification: DataClassification = DataClassification.CONFIDENTIAL,
        context: Optional[Dict[str, Any]] = None
    ) -> Tuple[str, Dict[str, Any]]:
        """
        Encrypt string and return base64-encoded ciphertext.

        Args:
            plaintext: String to encrypt
            classification: Data classification level
            context: Optional context metadata

        Returns:
            Tuple of (base64_ciphertext, metadata)
        """
        plaintext_bytes = plaintext.encode('utf-8')
        ciphertext, metadata = self.encrypt_data(plaintext_bytes, classification, context)
        ciphertext_b64 = base64.b64encode(ciphertext).decode('utf-8')

        return ciphertext_b64, metadata

    def decrypt_string(
        self,
        ciphertext_b64: str,
        metadata: Dict[str, Any]
    ) -> str:
        """
        Decrypt base64-encoded ciphertext to string.

        Args:
            ciphertext_b64: Base64-encoded ciphertext
            metadata: Encryption metadata

        Returns:
            Decrypted string
        """
        ciphertext = base64.b64decode(ciphertext_b64)
        plaintext_bytes = self.decrypt_data(ciphertext, metadata)
        plaintext = plaintext_bytes.decode('utf-8')

        return plaintext

    def _build_associated_data(
        self,
        classification: DataClassification,
        context: Optional[Dict[str, Any]]
    ) -> bytes:
        """Build associated data for AEAD."""
        aad = {
            'classification': classification.value,
            'context': context or {}
        }

        # Serialize to bytes
        import json
        aad_json = json.dumps(aad, sort_keys=True)
        return aad_json.encode('utf-8')

    def _validate_metadata(self, metadata: Dict[str, Any]) -> None:
        """
        Validate encryption metadata.

        Raises:
            DecryptionError: If metadata is invalid
        """
        required_fields = [
            'version', 'algorithm', 'key_id', 'nonce',
            'salt', 'kdf_iterations', 'classification'
        ]

        for field in required_fields:
            if field not in metadata:
                raise DecryptionError(f"Missing required metadata field: {field}")

        # Validate algorithm
        if metadata['algorithm'] != 'AES-256-GCM':
            raise DecryptionError(f"Unsupported algorithm: {metadata['algorithm']}")

        # Validate key ID
        if metadata['key_id'] != self.cipher.key_id:
            raise DecryptionError(
                f"Key mismatch: expected {self.cipher.key_id}, "
                f"got {metadata['key_id']}"
            )


# ============================================================================
# REDIS ENCRYPTION WRAPPER
# ============================================================================

class EncryptedRedisClient:
    """
    Redis client wrapper with transparent encryption.

    ISO 27001 A.10.1.1 - Automatic encryption of sensitive data
    HIPAA § 164.312(a)(2)(iv) - PHI encryption at rest

    Usage:
        >>> client = EncryptedRedisClient(redis_client, encryption_service)
        >>> client.set_encrypted("user:123:ssn", "123-45-6789", classification=DataClassification.PII)
        >>> value = client.get_decrypted("user:123:ssn")
    """

    def __init__(self, redis_client, encryption_service: EncryptionService):
        """
        Initialize encrypted Redis client.

        Args:
            redis_client: Redis client instance
            encryption_service: Encryption service
        """
        self.redis = redis_client
        self.encryption = encryption_service

        logger.info("Encrypted Redis client initialized")

    def set_encrypted(
        self,
        key: str,
        value: str,
        classification: DataClassification = DataClassification.CONFIDENTIAL,
        ttl: Optional[int] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> bool:
        """
        Set encrypted value in Redis.

        Args:
            key: Redis key
            value: Plaintext value to encrypt
            classification: Data classification level
            ttl: Optional TTL in seconds
            context: Optional context metadata

        Returns:
            True if successful
        """
        try:
            # Encrypt value
            ciphertext_b64, metadata = self.encryption.encrypt_string(
                plaintext=value,
                classification=classification,
                context=context
            )

            # Store encrypted data and metadata
            import json
            encrypted_package = {
                'ciphertext': ciphertext_b64,
                'metadata': metadata
            }

            encrypted_json = json.dumps(encrypted_package)

            # Store in Redis
            if ttl:
                self.redis.setex(key, ttl, encrypted_json)
            else:
                self.redis.set(key, encrypted_json)

            logger.debug(
                f"Encrypted value stored in Redis",
                extra={
                    'key': key,
                    'classification': classification.value,
                    'ttl': ttl
                }
            )

            return True

        except Exception as e:
            logger.error(f"Failed to store encrypted value: {e}", exc_info=True)
            return False

    def get_decrypted(self, key: str) -> Optional[str]:
        """
        Get and decrypt value from Redis.

        Args:
            key: Redis key

        Returns:
            Decrypted value or None if key doesn't exist

        Raises:
            DecryptionError: If decryption fails
        """
        try:
            # Get encrypted package from Redis
            encrypted_json = self.redis.get(key)

            if encrypted_json is None:
                return None

            # Parse package
            import json
            encrypted_package = json.loads(encrypted_json)

            ciphertext_b64 = encrypted_package['ciphertext']
            metadata = encrypted_package['metadata']

            # Decrypt
            plaintext = self.encryption.decrypt_string(ciphertext_b64, metadata)

            logger.debug(
                f"Decrypted value retrieved from Redis",
                extra={'key': key}
            )

            return plaintext

        except Exception as e:
            logger.error(f"Failed to decrypt value: {e}", exc_info=True)
            raise DecryptionError(f"Failed to decrypt value: {e}")

    def delete(self, key: str) -> bool:
        """Delete key from Redis."""
        result = self.redis.delete(key)
        return bool(result)

    def exists(self, key: str) -> bool:
        """Check if key exists in Redis."""
        return bool(self.redis.exists(key))


# ============================================================================
# FACTORY FUNCTIONS
# ============================================================================

def create_encryption_service(
    master_key: Optional[str] = None,
    kdf_iterations: Optional[int] = None
) -> EncryptionService:
    """
    Create encryption service from environment configuration.

    Args:
        master_key: Optional master key (defaults to ENCRYPTION_MASTER_KEY env var)
        kdf_iterations: Optional KDF iterations (defaults to KDF_ITERATIONS env var)

    Returns:
        EncryptionService instance

    Raises:
        ValueError: If master key is not configured
    """
    from app.core.config import get_settings

    settings = get_settings()

    # Get master key
    if master_key is None:
        master_key = os.getenv('ENCRYPTION_MASTER_KEY', '')
        if not master_key:
            raise ValueError(
                "ENCRYPTION_MASTER_KEY not configured. "
                "Set in environment or pass as argument."
            )

    # Get KDF iterations
    if kdf_iterations is None:
        kdf_iterations = getattr(settings, 'KDF_ITERATIONS', KeyDerivation.DEFAULT_ITERATIONS)

    return EncryptionService(
        master_key=master_key,
        kdf_iterations=kdf_iterations
    )


def create_encrypted_redis_client(redis_client) -> EncryptedRedisClient:
    """
    Create encrypted Redis client.

    Args:
        redis_client: Redis client instance

    Returns:
        EncryptedRedisClient instance
    """
    encryption_service = create_encryption_service()

    return EncryptedRedisClient(
        redis_client=redis_client,
        encryption_service=encryption_service
    )

File: frontend/src/utils/encryption.ts
/**
 * Data-at-Rest Encryption for IndexedDB
 * ISO 27001 A.10.1.1 - Policy on the use of cryptographic controls
 * ISO 27001 A.10.1.2 - Key management
 *
 * Provides AES-GCM encryption for sensitive data stored in browser IndexedDB.
 * Uses Web Crypto API for FIPS 140-2 compliant encryption.
 *
 * @module utils/encryption
 */

/**
 * Data classification levels (ISO 27001 A.8.2.1)
 */
export enum DataClassification {
  PUBLIC = 'public',
  INTERNAL = 'internal',
  CONFIDENTIAL = 'confidential',
  PHI = 'phi', // HIPAA Protected Health Information
  PII = 'pii', // Personal Identifiable Information
}

/**
 * Encryption metadata stored alongside ciphertext
 */
export interface EncryptionMetadata {
  version: string;
  algorithm: string;
  keyId: string;
  iv: string; // Base64-encoded initialization vector
  salt: string; // Base64-encoded salt
  classification: DataClassification;
  encryptedAt: string; // ISO 8601 timestamp
  context?: Record<string, any>;
}

/**
 * Encrypted package containing ciphertext and metadata
 */
export interface EncryptedPackage {
  ciphertext: string; // Base64-encoded
  metadata: EncryptionMetadata;
}

/**
 * Encryption configuration
 */
interface EncryptionConfig {
  algorithm: string;
  keySize: number;
  ivSize: number;
  saltSize: number;
  kdfIterations: number;
}

/**
 * Default encryption configuration
 * - AES-GCM: Authenticated encryption with associated data (AEAD)
 * - 256-bit keys: Maximum security
 * - 96-bit IV: Recommended for GCM mode
 * - 128-bit salt: NIST recommendation
 * - 100,000 iterations: NIST SP 800-132 recommendation
 */
const DEFAULT_CONFIG: EncryptionConfig = {
  algorithm: 'AES-GCM',
  keySize: 256,
  ivSize: 96,
  saltSize: 128,
  kdfIterations: 100000,
};

/**
 * AES-GCM Encryption Service
 *
 * Provides browser-based encryption using Web Crypto API.
 * All cryptographic operations use browser-native implementations.
 */
export class EncryptionService {
  private config: EncryptionConfig;
  private masterKey: string;
  private derivedKey: CryptoKey | null = null;
  private salt: Uint8Array | null = null;
  private keyId: string = '';

  constructor(masterKey: string, config: Partial<EncryptionConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.masterKey = masterKey;
  }

  /**
   * Initialize encryption service by deriving encryption key
   */
  async initialize(): Promise<void> {
    try {
      // Generate random salt
      this.salt = crypto.getRandomValues(new Uint8Array(this.config.saltSize / 8));

      // Derive key from master key using PBKDF2
      this.derivedKey = await this.deriveKey(this.masterKey, this.salt);

      // Calculate key fingerprint
      this.keyId = await this.calculateKeyId(this.derivedKey);

      console.debug('[Encryption] Service initialized', {
        keyId: this.keyId,
        algorithm: this.config.algorithm,
      });
    } catch (error) {
      console.error('[Encryption] Failed to initialize:', error);
      throw new Error(`Encryption service initialization failed: ${error}`);
    }
  }

  /**
   * Encrypt data with AES-GCM
   *
   * @param data - Plaintext data to encrypt
   * @param classification - Data classification level
   * @param context - Optional context metadata
   * @returns Encrypted package with metadata
   */
  async encrypt(
    data: string | ArrayBuffer,
    classification: DataClassification = DataClassification.CONFIDENTIAL,
    context?: Record<string, any>
  ): Promise<EncryptedPackage> {
    if (!this.derivedKey || !this.salt) {
      throw new Error('Encryption service not initialized. Call initialize() first.');
    }

    try {
      // Convert data to ArrayBuffer
      const plaintext = typeof data === 'string'
        ? new TextEncoder().encode(data)
        : new Uint8Array(data);

      // Generate random IV (MUST be unique for each encryption)
      const iv = crypto.getRandomValues(new Uint8Array(this.config.ivSize / 8));

      // Prepare additional authenticated data (AAD)
      const aad = this.buildAAD(classification, context);

      // Encrypt with AES-GCM
      const ciphertext = await crypto.subtle.encrypt(
        {
          name: this.config.algorithm,
          iv: iv as BufferSource,
          additionalData: aad as BufferSource,
        },
        this.derivedKey,
        plaintext
      );

      // Build metadata
      const metadata: EncryptionMetadata = {
        version: '1.0',
        algorithm: this.config.algorithm,
        keyId: this.keyId,
        iv: this.arrayBufferToBase64(iv),
        salt: this.arrayBufferToBase64(this.salt),
        classification,
        encryptedAt: new Date().toISOString(),
        context: context || {},
      };

      // Build encrypted package
      const package_: EncryptedPackage = {
        ciphertext: this.arrayBufferToBase64(ciphertext),
        metadata,
      };

      console.debug('[Encryption] Data encrypted', {
        keyId: this.keyId,
        classification,
        size: ciphertext.byteLength,
      });

      return package_;
    } catch (error) {
      console.error('[Encryption] Encryption failed:', error);
      throw new Error(`Encryption failed: ${error}`);
    }
  }

  /**
   * Decrypt data using stored metadata
   *
   * @param encryptedPackage - Encrypted package from encrypt()
   * @returns Decrypted plaintext
   */
  async decrypt(encryptedPackage: EncryptedPackage): Promise<string> {
    if (!this.derivedKey) {
      throw new Error('Encryption service not initialized. Call initialize() first.');
    }

    try {
      const { ciphertext, metadata } = encryptedPackage;

      // Validate metadata
      this.validateMetadata(metadata);

      // Convert ciphertext and IV from base64
      const ciphertextBuffer = this.base64ToArrayBuffer(ciphertext);
      const iv = this.base64ToArrayBuffer(metadata.iv);

      // Rebuild AAD
      const aad = this.buildAAD(
        metadata.classification as DataClassification,
        metadata.context
      );

      // Decrypt with AES-GCM (includes authentication tag verification)
      const plaintextBuffer = await crypto.subtle.decrypt(
        {
          name: this.config.algorithm,
          iv: iv as BufferSource,
          additionalData: aad as BufferSource,
        },
        this.derivedKey,
        ciphertextBuffer as BufferSource
      );

      // Convert to string
      const plaintext = new TextDecoder().decode(plaintextBuffer);

      console.debug('[Encryption] Data decrypted', {
        keyId: metadata.keyId,
        classification: metadata.classification,
      });

      return plaintext;
    } catch (error) {
      // Authentication tag verification failed - data was tampered with
      if (error instanceof DOMException && error.name === 'OperationError') {
        console.error('[Encryption] Authentication failed - data integrity violation');
        throw new Error('Data integrity violation: Authentication failed');
      }

      console.error('[Encryption] Decryption failed:', error);
      throw new Error(`Decryption failed: ${error}`);
    }
  }

  /**
   * Derive encryption key from master key using PBKDF2
   */
  private async deriveKey(
    masterKey: string,
    salt: Uint8Array
  ): Promise<CryptoKey> {
    // Import master key
    const keyMaterial = await crypto.subtle.importKey(
      'raw',
      new TextEncoder().encode(masterKey),
      { name: 'PBKDF2' },
      false,
      ['deriveBits', 'deriveKey']
    );

    // Derive key using PBKDF2
    const derivedKey = await crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt: salt as BufferSource,
        iterations: this.config.kdfIterations,
        hash: 'SHA-256',
      },
      keyMaterial,
      {
        name: this.config.algorithm,
        length: this.config.keySize,
      },
      false, // Not extractable
      ['encrypt', 'decrypt']
    );

    return derivedKey;
  }

  /**
   * Calculate key fingerprint (SHA-256 hash)
   */
  private async calculateKeyId(key: CryptoKey): Promise<string> {
    // Export key to get raw bytes
    const exportedKey = await crypto.subtle.exportKey('raw', key);

    // Hash with SHA-256
    const hashBuffer = await crypto.subtle.digest('SHA-256', exportedKey);

    // Convert to hex string (first 16 chars)
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');

    return hashHex.substring(0, 16);
  }

  /**
   * Build Additional Authenticated Data (AAD) for AEAD
   */
  private buildAAD(
    classification: DataClassification,
    context?: Record<string, any>
  ): Uint8Array {
    const aad = {
      classification,
      context: context || {},
    };

    const aadJson = JSON.stringify(aad);
    return new TextEncoder().encode(aadJson);
  }

  /**
   * Validate encryption metadata
   */
  private validateMetadata(metadata: EncryptionMetadata): void {
    const requiredFields: (keyof EncryptionMetadata)[] = [
      'version',
      'algorithm',
      'keyId',
      'iv',
      'salt',
      'classification',
    ];

    for (const field of requiredFields) {
      if (!(field in metadata)) {
        throw new Error(`Missing required metadata field: ${field}`);
      }
    }

    // Validate algorithm
    if (metadata.algorithm !== this.config.algorithm) {
      throw new Error(`Unsupported algorithm: ${metadata.algorithm}`);
    }

    // Validate key ID
    if (metadata.keyId !== this.keyId) {
      throw new Error(
        `Key mismatch: expected ${this.keyId}, got ${metadata.keyId}`
      );
    }
  }

  /**
   * Convert ArrayBuffer to base64 string
   */
  private arrayBufferToBase64(buffer: ArrayBuffer | Uint8Array): string {
    const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }

  /**
   * Convert base64 string to ArrayBuffer
   */
  private base64ToArrayBuffer(base64: string): Uint8Array {
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }
}

/**
 * Encrypted IndexedDB Storage
 *
 * Wrapper around IndexedDB with transparent encryption/decryption.
 * ISO 27001 A.10.1.1 - Automatic encryption of sensitive data
 */
export class EncryptedStorage {
  private dbName: string;
  private storeName: string;
  private encryption: EncryptionService;
  private db: IDBDatabase | null = null;

  constructor(
    dbName: string,
    storeName: string,
    encryptionService: EncryptionService
  ) {
    this.dbName = dbName;
    this.storeName = storeName;
    this.encryption = encryptionService;
  }

  /**
   * Open IndexedDB database
   */
  async open(): Promise<void> {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open(this.dbName, 1);

      request.onerror = () => {
        console.error('[EncryptedStorage] Failed to open database:', request.error);
        reject(new Error(`Failed to open database: ${request.error}`));
      };

      request.onsuccess = () => {
        this.db = request.result;
        console.debug('[EncryptedStorage] Database opened', {
          dbName: this.dbName,
          storeName: this.storeName,
        });
        resolve();
      };

      request.onupgradeneeded = (event) => {
        const db = (event.target as IDBOpenDBRequest).result;

        // Create object store if it doesn't exist
        if (!db.objectStoreNames.contains(this.storeName)) {
          db.createObjectStore(this.storeName);
          console.debug('[EncryptedStorage] Object store created', {
            storeName: this.storeName,
          });
        }
      };
    });
  }

  /**
   * Set encrypted value in IndexedDB
   *
   * @param key - Storage key
   * @param value - Plaintext value to encrypt
   * @param classification - Data classification level
   * @param context - Optional context metadata
   */
  async setEncrypted(
    key: string,
    value: string,
    classification: DataClassification = DataClassification.CONFIDENTIAL,
    context?: Record<string, any>
  ): Promise<void> {
    if (!this.db) {
      throw new Error('Database not opened. Call open() first.');
    }

    try {
      // Encrypt value
      const encryptedPackage = await this.encryption.encrypt(
        value,
        classification,
        context
      );

      // Store encrypted package
      return new Promise((resolve, reject) => {
        const transaction = this.db!.transaction([this.storeName], 'readwrite');
        const store = transaction.objectStore(this.storeName);
        const request = store.put(encryptedPackage, key);

        request.onerror = () => {
          console.error('[EncryptedStorage] Failed to store encrypted value:', request.error);
          reject(new Error(`Failed to store encrypted value: ${request.error}`));
        };

        request.onsuccess = () => {
          console.debug('[EncryptedStorage] Encrypted value stored', {
            key,
            classification,
          });
          resolve();
        };
      });
    } catch (error) {
      console.error('[EncryptedStorage] Failed to set encrypted value:', error);
      throw error;
    }
  }

  /**
   * Get and decrypt value from IndexedDB
   *
   * @param key - Storage key
   * @returns Decrypted value or null if not found
   */
  async getDecrypted(key: string): Promise<string | null> {
    if (!this.db) {
      throw new Error('Database not opened. Call open() first.');
    }

    try {
      const encryptedPackage = await new Promise<EncryptedPackage | null>(
        (resolve, reject) => {
          const transaction = this.db!.transaction([this.storeName], 'readonly');
          const store = transaction.objectStore(this.storeName);
          const request = store.get(key);

          request.onerror = () => {
            console.error('[EncryptedStorage] Failed to retrieve value:', request.error);
            reject(new Error(`Failed to retrieve value: ${request.error}`));
          };

          request.onsuccess = () => {
            resolve(request.result || null);
          };
        }
      );

      if (!encryptedPackage) {
        return null;
      }

      // Decrypt value
      const plaintext = await this.encryption.decrypt(encryptedPackage);

      console.debug('[EncryptedStorage] Decrypted value retrieved', { key });

      return plaintext;
    } catch (error) {
      console.error('[EncryptedStorage] Failed to get decrypted value:', error);
      throw error;
    }
  }

  /**
   * Delete key from IndexedDB
   */
  async delete(key: string): Promise<void> {
    if (!this.db) {
      throw new Error('Database not opened. Call open() first.');
    }

    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readwrite');
      const store = transaction.objectStore(this.storeName);
      const request = store.delete(key);

      request.onerror = () => {
        console.error('[EncryptedStorage] Failed to delete key:', request.error);
        reject(new Error(`Failed to delete key: ${request.error}`));
      };

      request.onsuccess = () => {
        console.debug('[EncryptedStorage] Key deleted', { key });
        resolve();
      };
    });
  }

  /**
   * Check if key exists in IndexedDB
   */
  async exists(key: string): Promise<boolean> {
    if (!this.db) {
      throw new Error('Database not opened. Call open() first.');
    }

    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readonly');
      const store = transaction.objectStore(this.storeName);
      const request = store.getKey(key);

      request.onerror = () => {
        console.error('[EncryptedStorage] Failed to check existence:', request.error);
        reject(new Error(`Failed to check existence: ${request.error}`));
      };

      request.onsuccess = () => {
        resolve(request.result !== undefined);
      };
    });
  }

  /**
   * Close database connection
   */
  close(): void {
    if (this.db) {
      this.db.close();
      this.db = null;
      console.debug('[EncryptedStorage] Database closed');
    }
  }
}

/**
 * Create encryption service with master key from environment
 */
export function createEncryptionService(): EncryptionService {
  // In production, this should come from a secure source (e.g., derived from user password)
  const masterKey = import.meta.env.VITE_ENCRYPTION_KEY || 'development-master-key-CHANGE-IN-PRODUCTION';

  if (masterKey === 'development-master-key-CHANGE-IN-PRODUCTION') {
    console.warn(
      '[Encryption] Using default development master key. ' +
      'Set VITE_ENCRYPTION_KEY in production!'
    );
  }

  return new EncryptionService(masterKey);
}

/**
 * Create encrypted IndexedDB storage
 */
export async function createEncryptedStorage(
  dbName: string = 'medical-imaging-viewer',
  storeName: string = 'encrypted-data'
): Promise<EncryptedStorage> {
  const encryption = createEncryptionService();
  await encryption.initialize();

  const storage = new EncryptedStorage(dbName, storeName, encryption);
  await storage.open();

  return storage;
}
